File: kotlin/kotlin_from_entry_to_advanced_combat/src/com/easykotlin/test/chapter6/ListUtilTest.java
package com.easykotlin.test.chapter6;

import java.util.ArrayList;
import java.util.List;

public class ListUtilTest<T> {
    public List<T> filter(List<T> list, Predicate<T> p) {
        List<T> result = new ArrayList<>();
        for (T t : list) {
            if (p.predicate(t))
                result.add(t);
        }

        return result;
    }
}

interface Predicate<T> {
    Boolean predicate(T t);
}

File: kotlin/kotlin_from_entry_to_advanced_combat/src/com/easykotlin/test/chapter6/ListUtilTest.kt
package com.easykotlin.test.chapter6

fun <T> List<T>.filters(predicate: (T) -> Boolean): MutableList<T> {
    val result = ArrayList<T>()
    this.forEach {
        if (predicate(it)) {
            result.add(it)
        }
    }

    return result
}

fun main(args: Array<String>) {
    val list = mutableListOf(1, 2, 3, 4, 5, 6, 7)
    println(list.filters { it -> it % 2 == 1 })
}